src/lib/datepicker/calendar.ts

```diff
@@ -346,7 +338,7 @@ export class MatCalendar<D> implements AfterContentInit, OnDestroy, OnChanges {
             this._dateAdapter.addCalendarMonths(this._activeDate, 1);
         break;
       case ENTER:
-        if (this._dateFilterForViews(this._activeDate)) {
+        if (this.dateFilter(this._activeDate)) {
```

Comment:
do we need to check the min and max here too? or will we never wind up in that situation?

Reply:
I'll check it more carefully. I've just replaced without considering this possibility.

Reply:
In fact I couldn't imagine a scenario where the user is able to select a date with keyboard.... as soon as I set max and min dates inside `<month-view>`, and I've just noticed that I forgot to do it. 😳 

One option would be to keep the `_dateFilterForViews` and don't pass `maxDate` and `minDate` to `<month-view>`, but it would be potentially confusing in the future (why only the filters of `month-view` would be in `calendar.ts`?) To keep it consistent, all of filtering logic should follow the same rule. So each view has all the filtering logic inside.

I can go on the other way if you think it's better: keep all the filtering logic in `calendar.ts` (and don't pass `maxDate` and `minDate` to the `views` components. Let me know if you prefer this option.

Reply:
Oh wait this is coming back to me now, I think what happened is that I allowed people to navigate to disabled dates with the keyboard for the following reason: imagine someone created a date filter that returns `date < 1/1/2018`, now if I try to arrow forward from 2017 and go to the next enabled date... my application hangs forever. So instead I just made it impossible to select the date if the active date is disabled.

So I think we do need to make sure we somehow check the min and max dates. We should actually be doing it for year and multi-year view too, I think it's a bug that we're currently not. I feel like the way this _should_ work is that the calendar should delegate the keyboard handling to the appropriate view rather than having the code for all of the views hardcoded in the calendar, but its already kind of set up weird, so however you want to make it work is fine...

Reply:
Moving each keyboard handler to its own view would be nice, but it would also add a delay to this PR, so what if we do it in a separate one? I think the performance issue that lead to this PR is really anoying and should be merged ASAP.

Reply:
The problem is the `ENTER` case for each of these handlers needs to check the whether the date is disabled in the corresponding view. So for this PR we need to do one of the following:
1. Wire things up so the calendar can delegate checking the disabled-ness of a date to the view (at this point we might as well just move the whole handler)
2. Move all of the disabled-ness checks into the calendar and have the views ask the calendar (kind of moving in the wrong direction)
3. Duplicate the disabled-ness checking logic (bleh)

Reply:
We have `@Input() allowDisabledSelection` in `mat-calendar-body`. It's used in `<year-view>` and in `<mat-year-view>`. We could see whether it can be used in `<month-view>` too, and get rid of the `ENTER` key filter logic... in a future PR.